### DISCUSSION TITLE
Accessing Parquet file directly from Azure Blob Container URL

### DISCUSSION THREAD

**Discussion by haidernitc** [duckdb/duckdb#2565] [Q&A]
I would like to understand whether we can directly access Parquet file from Azure Blob Container URL instead of downloading it locally.

Example:
```
 select * from parquet_schema("https://abc.blob.core.windows.net/folder1/partition_folder/part-00000-c000.snappy.parquet")
```

**Comment by asbisen** [comment]
You can access parquet stored on Azure Blob using pyarrow. The performance seems to be pretty decent.

```python
from adlfs import AzureBlobFileSystem
import duckdb
import pyarrow.parquet as pq
import pyarrow.dataset as ds


account_name = "xxxxx"
account_key = "xxxxx"
abfs = AzureBlobFileSystem( account_name = account_name, account_key = account_key, container_name = "data")
pqdata = ds.dataset("path/inside/abfs", filesystem=abfs)


conn = duckdb.connect(":memory:")
conn.execute("SELECT * from pqdata LIMIT 10")





```

**Reply by astrowonk** [reply]
This sort of works, but even if you pass a list of specific blob names to ABFS (not an entire container), ABFS will try to `ls` the entire container. If the container has a lot of blobs, the initial query such as above will take a long time (a minute). I'm not sure if duckb is invoking this `ls` command or something inside the pyarrow dataset object but it adds some startup time.

**Reply by IamJeffG** [reply]
I also don't see partition filters being used at query time:

```
account_name = "xxxxx"
account_key = "xxxxx"
abfs = AzureBlobFileSystem( account_name = account_name, account_key = account_key, container_name = "data")
partitioning = ds.HivePartitioning(pa.schema([("segment", pa.string()), ("ingest_date", pa.string())]))
pqdata = ds.dataset("path/inside/abfs", filesystem=abfs, partitioning=partitioning)

conn = duckdb.connect(":memory:")
conn.execute("SELECT * from pqdata WHERE segment='abcd'")
```

Good: This solution correctly parses and returns the `segment` in the blob path as a bona-fide column.

Bad: the query engine does not know that it only needs to scan a subset of the blobs (those prefixed with `path/inside/abfs/segment=abcd/`).  It scans all blobs having all `path/inside/abfs/segment=abcd/segment=` prefixes.  Which, in turn, pulls back way too much data to my machine.

**Comment by jboarman** [comment]
Ideally, a more general solution would be to provide support for [fsspec](https://filesystem-spec.readthedocs.io/en/latest/).  However, perhaps since duckdb's core code is written in C++ then there is a bit more work needed to generalize support for additional file systems outside s3?

**Reply by Mause** [reply]
`fsspec` is now supported in DuckDB's Python library :grin:

**Reply by cboettig** [reply]
@Mause that's awesome! Can you point us to a simple example of doing this?  

(e.g. Microsoft PlanetaryComputer has various [example notebooks](https://planetarycomputer.microsoft.com/dataset/gbif#Example-Notebook) showing access to parquet files on azure block storage, but that example shows rather cumbersome approach with dask that only works on a single parquet partition file at a time.)

**Reply by djouallah** [reply]
self promotion :)
https://datamonkeysite.com/2023/01/27/querying-azure-storage-using-duckdb/

**Reply by Mause** [reply]
Additional docs are available here: https://duckdb.org/docs/guides/python/filesystems

**Reply by cboettig** [reply]
thanks, love it!

**Reply by astrowonk** [reply]
Fsspec support is great, but ABFS still has problems with containers with lots of blobs. 

this is pretty fast:

```python


df = con.execute(f"""
    select * mean(value),thedate
    from read_parquet('abfs://my-container/my-specific-file.parquet') group by thedate
    """
).df()

```

is pretty fast (and nifty this works!). But 

```python

df = con.execute(f"""
    select * mean(value),thedate
    from read_parquet('abfs://my-container/my-file-prefix 2022*.parquet') group by thedate
    """
).df()

```

is quite slow. Even the _second_ run is slow, implying that there's no caching of the file names of the blobs in the container. There is a long delay  before the cool tqdm-like progress bar pops up.

The good news is that read_parquet can take a [] of urls! and that seems pretty fast. I built my own list of 60 azure blob urls and that scanned and loaded in 5 seconds!

I'll post something about this in the adlfs repository. I understand the first scan for a wildcard should be long but not sure why the second one is.

**Reply by cboettig** [reply]
Agreed, wrapping of adlfs seems to time out for me on non-trivial queries.  Here's a quick example querying GBIF, a large-ish parquet database that is publicly available on both AWS S3 and Azure: https://github.com/cboettig/example-notebooks/blob/main/gbif-duckdb.ipynb.  (good duckdb vibes, the most observed species in this dataset is a :duck:).  The S3 example seems to run fine (using duckdb's httpfs module, so no fspec wrapping involved), but the same query times out on Azure.

**Reply by astrowonk** [reply]
Any idea how many files are in that container? My guess is the * wildcard leads to a scan of the entire container and that's slow. Though in my own experience it's never timing out persay, it's just taking ~2 min.

The way we build our parquet files there is a mongo collection with metadata on them, so I was able to make a wrapper class that lets me query `DUCKDB` where the `DUCKDB` string is replaced by a long programmatically generated construct of `['abfs://blob1','abfs://blob2']` etc. That's leading to some very quick group bys of like 2 seconds for 30 blobs. Not too shabby.

**Reply by cboettig** [reply]
@astrowonk a quick ls* on the bucket (`gbif-open-data-us-east-1/occurrence/2023-02-01/occurrence.parquet/`) takes about 2 seconds on my machine and shows there are 2,053 parquet files.  Hive partitioning is of course a good way to improve performance when the data structure lends itself to that, and I think is already supported in duckdb now, but I don't think the GBIF data has any natural way to create even partitions.  in any event the wildcard read seems quite fast with duckdb `httpfs` -- the bottleneck seems to be the python wrapping in `fsspec` / `adlfs` I think?  

If I run the azure example from an Azure machine, the adlf approach works fine without timing out (and as expected, is faster than accessing the AWS-hosted data from outside the data center / region).  but it breaks for me on my local machine while the `httpfs` mechansim is just fine.

**Reply by astrowonk** [reply]
I take it back. Even hard coding a list of urls (or one url) and AzureBlobFileSystem scans the entire container unnecessarily.

**Comment by carlopi** [comment]
Have you tried https://github.com/duckdblabs/duckdb_azure?

**Reply by astrowonk** [reply]
oh wow, no! I'll try this on Monday and test it out, thank you!

**Reply by SzilvasiPeter** [reply]
The Azure extension uses a storage connection string for authentication. Is it possible to use a **SAS token** instead of a connection string?

**Reply by erik-farmer** [reply]
@SzilvasiPeter Did you ever get a SAS token implementation working?

**Reply by SzilvasiPeter** [reply]
@erik-farmer Yes, it is working. I did not even install the Azure extension when using SAS token. I have just downloaded the `duckdb` binary, executed it. After that, I generated the SAS token on a dummy file and imported in the following way:

```sql
D SELECT * FROM 'https://<storage_account>.blob.core.windows.net/data/data.csv?<SAS_token>';
┌──────────┬───────┬────────┐
│   name   │  age  │ height │
│ varchar  │ int64 │ int64  │
├──────────┼───────┼────────┤
│ John Doe │    12 │    165 │
│ Jane Doe │    13 │    150 │
└──────────┴───────┴────────┘
```
